lib/vanagon/component/source/git.rb

```diff
@@ -34,30 +35,20 @@ class << self
           # with: NoMethodError: undefined method `split' for nil:NilClass
           #
           # We'll work around that case by calling 'git ls-remote' directly ourselves.
-          #
-          # I'm leaving in the broken version here for a time when the ruby-git library
-          # is fixed.
-
-          #def valid_remote?(url, timeout = 0)
-          #  Timeout.timeout(timeout) do
-          #    !!::Git.ls_remote(url)
-          #  end
-          #rescue ::Git::GitExecuteError
-          #  false
-          #rescue Timeout::Error
-          #  false
-          #end
 
           def valid_remote?(url, timeout = 0)
-            Timeout.timeout(timeout) do
-              Vanagon::Utilities.local_command("git ls-remote --heads #{url} > /dev/null 2>&1")
-              return false unless $?.exitstatus.zero?
-              return true
+            # RE-15209. To relieve github rate-limiting, if the URL starts with
+            # https://github.com/... just accept it rather than ping github over and over.
+            return true if url.to_s.start_with?('https://github.com/')
+
+            begin
+              Timeout.timeout(timeout) do
+                Vanagon::Utilities.local_command("git ls-remote --heads #{url} > /dev/null 2>&1")
+                return $?.exitstatus.zero?
+              end
+            rescue StandardError
+              return false
             end
-          rescue Timeout::Error
-            return false
-          rescue RuntimeError
-            return false
```

Comment:
The `valid_remote?` method will now rescue `StandardError` (or derived classes) that aren't `RuntimeError`. I think it only matters if callers are expecting the build to fail if one of those other exceptions are raised like `ArgumentError`?  This picture is helpful https://www.honeybadger.io/blog/understanding-the-ruby-exception-hierarchy/

Reply:
Yup, that was too aggressive. Remedy could be putting it back:

```ruby
rescue Timeout::Error
  return false
rescue RuntimeError
  return false
```
but this reads as misleading to me because I take this as an effective shadow and it smells a little bit.

But, getting rid of the shadow:

```ruby
rescue RuntimeError
  return false
```
relieves the smell but then explicitly loses the reason for the failure, Timeout vs. git ls-remote. 

I'm actually leaning towards the former. Any thoughts?

Reply:
> because I take this as an effective shadow

As long as the more specific/derived class is rescued first, which it is in this case, then it's ok. That pattern is often used when you need different behavior based on the exception class:

```ruby
rescue A => a
  puts a.method1
rescue B => b
  puts b.method2
end
```
 
However, in this case, `Timeout::Error` extends `RuntimeError` and we return `false` in both cases, so there's no real benefit to rescuing them separately.